### PR TITLE
* add install_on_device action

### PR DESCRIPTION
**Pull request by hjanuschka**
adds a action `install_on_device`

lets you deploy an ipa to any connected device, using ios-deploy under the hood.

because we need some extra work to get an app to device since `match`. i thought deploying via commandline would be nice.

demo (truncated):

```
[21:08:49]: Driving the lane 'dd' 🚀
[21:08:49]: ------------------
[21:08:49]: --- Step: taxi ---
[21:08:49]: ------------------
[21:08:49]: $ ios-deploy  --bundle fivexfive.ipa
[21:08:49]: ▸ [....] Waiting for iOS device to be connected
[21:08:49]: ▸ [....] Using iPhone 6s Plus 'Helmut Januschka’s iPhone4 (2)' (a3be6c9ff7e5c3c6028597513243b0f933b876d4).
[21:08:49]: ▸ ------ Install phase ------
[21:08:49]: ▸ [  0%] Found iPhone 6s Plus 'Helmut Januschka’s iPhone4 (2)' (a3be6c9ff7e5c3c6028597513243b0f933b876d4) connected through USB, beginning install
[21:08:49]: ▸ [  5%] Copying META-INF/ to device
[21:08:49]: ▸ [  5%] Copying META-INF/com.apple.ZipMetadata.plist to device
[21:08:49]: ▸ [  5%] Copying Payload/ to device
[21:08:50]: ▸ [ 49%] Copying Symbols/ to device
[21:08:50]: ▸ [ 49%] Copying Symbols/4E2E2262-74E7-34E9-8431-9AE472D239DE.symbols to device
[21:08:50]: ▸ [ 49%] Copying Symbols/5F1CA885-E477-3545-9B84-7B26E96C37BA.symbols to device
[21:08:51]: ▸ [ 52%] CreatingStagingDirectory
[21:08:51]: ▸ [ 57%] ExtractingPackage
[21:08:51]: ▸ [ 60%] InspectingPackage
[21:08:51]: ▸ [ 60%] TakingInstallLock
[21:08:51]: ▸ [ 65%] PreflightingApplication
[21:08:51]: ▸ [ 65%] InstallingEmbeddedProfile
[21:08:51]: ▸ [ 70%] VerifyingApplication
[21:08:51]: ▸ [ 75%] CreatingContainer
[21:08:51]: ▸ [ 80%] InstallingApplication
[21:08:51]: ▸ [ 85%] PostflightingApplication
[21:08:51]: ▸ [ 90%] SandboxingApplication
[21:08:51]: ▸ [ 95%] GeneratingApplicationMap
[21:08:51]: ▸ [100%] Installed package fivexfive.ipa
[21:08:51]: Deployd fivexfive.ipa to device!

+------+-----------------+-------------+
|           fastlane summary           |
+------+-----------------+-------------+
| Step | Action          | Time (in s) |
+------+-----------------+-------------+
| 1    | skip_docs       | 0           |
| 2    | import_from_git | 0           |
| 3    | install_on_device            | 2           |
```

### CODE REVIEW

File: lib/fastlane/actions/taxi.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+module Fastlane
+  module Actions
+    module SharedValues
+    end
+
+    class TaxiAction < Action
+
+      def self.run(params)
+        taxi_cmd = [
+          "ios-deploy",
```
Comment (KrauseFx) [inline lib/fastlane/actions/taxi.rb]: There should be some error handling if that's not installed.

File: lib/fastlane/actions/taxi.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
+
+module Fastlane
+  module Actions
+    module SharedValues
+    end
+
+    class TaxiAction < Action
```
Comment (KrauseFx) [inline lib/fastlane/actions/taxi.rb]: I think we should name it more of what it's actually doing. How about `install_on_device`?

File: lib/fastlane/actions/taxi.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+          params[:extra],
+          "--bundle",
+          params[:ipa]
+
+        ]
+
+        taxi_cmd << "--no-wifi" if params[:skip_wifi]
+        taxi_cmd << ["--id", params[:device_id]] if params[:device_id]
+        Actions.sh(taxi_cmd.join(" "))
+        Helper.log.info "Deployd #{params[:ipa]} to device!"
```
Comment (KrauseFx) [inline lib/fastlane/actions/taxi.rb]: Spelling mistake here

File: lib/fastlane/actions/taxi.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                                      ),
+          FastlaneCore::ConfigItem.new(key: :ipa,
+                                       short_option: "-i",
+                                       env_name: "FL_TAXI_IPA",
+                                       description: "The IPA file to put on the device",
+                                       optional: true,
+                                       is_string: true,
+                                       default_value: Dir["*.ipa"].first,
+                                       verify_block: proc do |value|
+                                                       raise "Could not find ipa file at path '#{value}'" unless File.exist? value
```
Comment (KrauseFx) [inline lib/fastlane/actions/taxi.rb]: Could you move these blocks to the left to be only 2 spaces intended from `verify_block`

File: docs/Actions.md [hunk 5]
```diff
@@ -743,6 +743,18 @@ increment_build_number({
...
+### Local Device
+
+```ruby
+taxi(
+  device_id: "a3be6c9ff7e5c3c6028597513243b0f933b876d4"
+  ipa: "./app.ipa"
+  )
+```
+
+Installs the ipa on the device, if no id is given, the first found ios device will be used, works via usb or wifi.
```
Comment (KrauseFx) [inline docs/Actions.md]: ios should be iOS

Comment (KrauseFx) [inline docs/Actions.md]: usb should be USB
wifi should be Wi-Fi

File: docs/Actions.md [hunk 6]
```diff
@@ -743,6 +743,18 @@ increment_build_number({
...
+this requires `ios-deploy` to be installed please have a look at [ios-deploy](https://github.com/phonegap/ios-deploy), in short it is a `npm -g i ios-deploy`
```
Comment (KrauseFx) [inline docs/Actions.md]: `in short it is a `
please change to 
`to quickly install it, use `

File: lib/fastlane/actions/taxi.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                                       optional: true,
+                                       is_string: false
+                                      ),
+          FastlaneCore::ConfigItem.new(key: :ipa,
+                                       short_option: "-i",
+                                       env_name: "FL_TAXI_IPA",
+                                       description: "The IPA file to put on the device",
+                                       optional: true,
+                                       is_string: true,
+                                       default_value: Dir["*.ipa"].first,
```
Comment (KrauseFx) [inline lib/fastlane/actions/taxi.rb]: Default value should be `Actions.lane_context[SharedValues::IPA_OUTPUT_PATH] || Dir["*.ipa"].first` to automatically use the path to the ipa from the `gym` action :+1:

**Comment by hjanuschka** [comment]
hey, thx for review

 * feedback implemented
 * action renamed to `install_on_device`
 * checks if ios-deploy is installed
 * added a simple rspec test
 * ipa path default value changed
 

@KrauseFx  hope this works out - thx for your time on taking notes!

**Comment by KrauseFx** [comment]
This is great, thanks for updating the PR. The tests are failing due to the description: https://travis-ci.org/fastlane/fastlane/jobs/105918338

File: docs/Actions.md [hunk 8]
```diff
@@ -743,6 +743,19 @@ increment_build_number({
...
+### `install_on_device`
+
+Installs the ipa on the device, if no id is given, the first found iOS device will be used, works via USB or Wi-Fi.
+this requires `ios-deploy` to be installed please have a look at [ios-deploy](https://github.com/phonegap/ios-deploy). to quickly install it, use `npm -g i ios-deploy`
+
+```ruby
+install_on_device(
+  device_id: "a3be6c9ff7e5c3c6028597513243b0f933b876d4"
+  ipa: "./app.ipa"
+  )
```
Comment (KrauseFx) [inline docs/Actions.md]: The `)` should be on the left

File: lib/fastlane/actions/install_on_device.rb [hunk 9]
```diff
@@ -0,0 +1,85 @@
+
+module Fastlane
+  module Actions
+    module SharedValues
+    end
+
+    class InstallOnDeviceAction < Action
+
```
Comment (KrauseFx) [inline lib/fastlane/actions/install_on_device.rb]: Please remove newline

File: lib/fastlane/actions/install_on_device.rb [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+      def self.run(params)
+        unless Helper.test?
+          raise "ios-deploy not installed, see https://github.com/phonegap/ios-deploy for instructions".red if `which ios-deploy`.length == 0
+        end
+        taxi_cmd = [
+          "ios-deploy",
+          params[:extra],
+          "--bundle",
+          params[:ipa]
+
```
Comment (KrauseFx) [inline lib/fastlane/actions/install_on_device.rb]: Please remove newline

**Comment by KrauseFx** [comment]
Thanks again for all your great contribution :+1:

**Comment by hjanuschka** [comment]
* newlines removed
 * current master rebased


@KrauseFx  thx for the hints.

**Comment by KrauseFx** [comment]
Sorry for the trouble, maybe you've just missed my message, but the tests are failing: https://travis-ci.org/fastlane/fastlane/jobs/106310007 This is due to the action description. You can run the tests locally using `bundle exec rspec` :+1:

**Comment by hjanuschka** [comment]
@KrauseFx  ohh sorry i totally missed the comment about the documentation fail.
sorry for that - tests are clean now, master is rebased happy :octopus:

**Comment by KrauseFx** [comment]
Perfect, thank you! Let's :rocket: it :sparkles:

**Pull request merged**